File: js/u64_dyn.js
"use strict";

function get_bias(byte_length) {
  let bias = 0n;
  while (--byte_length) {
    bias += 1n;
    bias <<= 7n;
  }
  return bias;
}

function pack_u64_dyn(v) {
  if (typeof v !== "bigint") v = BigInt(v);
  const out = new Uint8Array(9);
  let length = 0;
  for (; length < 8; length++) {
    const cur = v & 0x7fn;
    v >>= 7n;
    if (v !== 0n) {
      out[length] = Number(cur | 0x80n);
    } else {
      out[length] = Number(cur);
      return out.subarray(0, length + 1);
    }
  }
  out[length++] = Number(v);
  return out.subarray(0, length);
}

function unpack_u64_dyn(buf, offset = 0) {
  let v = 0n;
  let bits = 0n;
  let used = 0;
  for (; used < 8; used++) {
    if (offset + used >= buf.length) throw new RangeError("Insufficient data");
    const b = BigInt(buf[offset + used]);
    v += (b & 0x7fn) << bits;
    if ((b & 0x80n) === 0n) {
      return [v, offset + used + 1];
    }
    bits += 7n;
  }
  if (offset + used >= buf.length) throw new RangeError("Insufficient data");
  const b = BigInt(buf[offset + used]);
  v += b << 56n;
  return [v, offset + used + 1];
}

function pack_u64_dyn_b(v) {
  if (typeof v !== "bigint") v = BigInt(v);
  const out = new Uint8Array(9);
  let length = 0;
  for (; length < 8; length++) {
    const cur = v & 0x7fn;
    v >>= 7n;
    if (v !== 0n) {
      out[length] = Number(cur | 0x80n);
      v -= 1n; // bias
    } else {
      out[length] = Number(cur);
      return out.subarray(0, length + 1);
    }
  }
  out[length++] = Number(v);
  return out.subarray(0, length);
}

function unpack_u64_dyn_b(buf, offset = 0) {
  let v = 0n;
  let bits = 0n;
  let used = 0;
  let bias = 0n;
  while (true) {
    if (offset + used >= buf.length) {
      throw new RangeError("Insufficient data");
    }
    const b = BigInt(buf[offset + used++]);
    if (used == 9) {
      v += b << 56n;
      break;
    }
    v += (b & 0x7fn) << bits;
    if ((b & 0x80n) === 0n) {
      break;
    }
    bits += 7n;
    bias += 1n << bits;
  }
  if (v > 0xffffffffffffffffn - bias) {
    throw new RangeError("Invalid data");
  }
  v += bias;
  return [v, offset + used];
}

function pack_u64_dyn_p(v) {
  if (typeof v !== "bigint") v = BigInt(v);
  const byte_length =
    1 +
    (v >= 128n) +
    (v >= 16384n) +
    (v >= 2097152n) +
    (v >= 268435456n) +
    (v >= 34359738368n) +
    (v >= 4398046511104n) +
    (v >= 562949953421312n) +
    (v >= 72057594037927936n);
  const first_byte_value_bits = (byte_length < 8) * (8 - byte_length);
  const first_byte_prefix_bits = byte_length - 1;
  const out = new Uint8Array(byte_length);
  out[0] =
    (0xff << (8 - first_byte_prefix_bits)) |
    (Number(v & 0xffn) & ((1 << first_byte_value_bits) - 1));
  v >>= BigInt(first_byte_value_bits);
  for (let i = 1; i != byte_length; ++i) {
    out[i] = Number((v >> BigInt((i - 1) * 8)) & 0xffn);
  }
  return out;
}

function unpack_u64_dyn_p(buf, offset = 0) {
  if (offset >= buf.length) {
    throw new RangeError("Insufficient data");
  }
  const byte_length = 1 + (Math.clz32(~buf[offset] & 0xff) - 24);
  const first_byte_value_bits = (byte_length < 8) * (8 - byte_length);
  if (offset + byte_length > buf.length) {
    throw new RangeError("Insufficient data");
  }
  let v = 0n;
  for (let i = 1; i != byte_length; ++i) {
    v |= BigInt(buf[offset + i]) << BigInt((i - 1) * 8);
  }
  v <<= BigInt(first_byte_value_bits);
  v |= BigInt(buf[offset] & ((1 << first_byte_value_bits) - 1));
  return [v, offset + byte_length];
}

function pack_u64_dyn_bp(v) {
  if (typeof v !== "bigint") v = BigInt(v);
  const byte_length =
    1 +
    (v >= 128n) +
    (v >= 16512n) +
    (v >= 2113664n) +
    (v >= 270549120n) +
    (v >= 34630287488n) +
    (v >= 4432676798592n) +
    (v >= 567382630219904n) +
    (v >= 72624976668147840n);
  const first_byte_value_bits = (byte_length < 8) * (8 - byte_length);
  const first_byte_prefix_bits = byte_length - 1;

  v -= get_bias(byte_length);

  const out = new Uint8Array(byte_length);
  out[0] =
    (0xff << (8 - first_byte_prefix_bits)) |
    (Number(v & 0xffn) & ((1 << first_byte_value_bits) - 1));
  v >>= BigInt(first_byte_value_bits);
  for (let i = 1; i != byte_length; ++i) {
    out[i] = Number((v >> BigInt((i - 1) * 8)) & 0xffn);
  }
  return out;
}

function unpack_u64_dyn_bp(buf, offset = 0) {
  if (offset >= buf.length) {
    throw new RangeError("Insufficient data");
  }
  const byte_length = 1 + (Math.clz32(~buf[offset] & 0xff) - 24);
  const first_byte_value_bits = (byte_length < 8) * (8 - byte_length);
  if (offset + byte_length > buf.length) {
    throw new RangeError("Insufficient data");
  }
  let v = 0n;
  for (let i = 1; i != byte_length; ++i) {
    v |= BigInt(buf[offset + i]) << BigInt((i - 1) * 8);
  }
  v <<= BigInt(first_byte_value_bits);
  v |= BigInt(buf[offset] & ((1 << first_byte_value_bits) - 1));

  const bias = get_bias(byte_length);
  if (v > 0xffffffffffffffffn - bias) {
    throw new RangeError("Invalid data");
  }
  return [v + bias, offset + byte_length];
}

function pack_i64_dyn_a(v) {
  if (typeof v !== "bigint") v = BigInt(v);
  const neg = v < 0n ? 1n : 0n;
  let u = v;
  if (neg) {
    u = (~v + 1n) & ~(1n << 63n);
  }
  const packed = (neg << 6n) | ((u & ~0x3fn) << 1n) | (u & 0x3fn);
  return pack_u64_dyn(packed);
}

function unpack_i64_dyn_a(buf, offset = 0) {
  const [u64, idx] = unpack_u64_dyn(buf, offset);
  let u = u64;
  const neg = (u >> 6n) & 1n;
  u = ((u >> 1n) & ~0x3fn) | (u & 0x3fn);
  let v = u;
  if (neg) {
    v = ~(u - 1n) | (1n << 63n);
    v = BigInt.asIntN(64, v);
  }
  return [v, idx];
}

function pack_i64_dyn_b(v) {
  if (typeof v !== "bigint") v = BigInt(v);
  const neg = v < 0n ? 1n : 0n;
  let u;
  if (neg) {
    u = ~v;
  } else {
    u = v;
  }
  const packed = (neg << 6n) | ((u & ~0x3fn) << 1n) | (u & 0x3fn);
  return pack_u64_dyn_b(packed);
}

function unpack_i64_dyn_b(buf, offset = 0) {
  const [u64, idx] = unpack_u64_dyn_b(buf, offset);
  let u = u64;
  const neg = (u >> 6n) & 1n;
  u = ((u >> 1n) & ~0x3fn) | (u & 0x3fn);
  let v;
  if (neg) {
    v = ~u;
  } else {
    v = u;
  }
  return [v, idx];
}

function pack_i64_dyn_bp(v) {
  if (typeof v !== "bigint") v = BigInt(v);
  const neg = v < 0n ? 1n : 0n;
  let u;
  if (neg) {
    u = ~v;
  } else {
    u = v;
  }
  const packed = (neg << 6n) | ((u & ~0x3fn) << 1n) | (u & 0x3fn);
  return pack_u64_dyn_bp(packed);
}

function unpack_i64_dyn_bp(buf, offset = 0) {
  const [u64, idx] = unpack_u64_dyn_bp(buf, offset);
  let u = u64;
  const neg = (u >> 6n) & 1n;
  u = ((u >> 1n) & ~0x3fn) | (u & 0x3fn);
  let v;
  if (neg) {
    v = ~u;
  } else {
    v = u;
  }
  return [v, idx];
}

function pack_u64_dyn_v2(v) {
  return pack_u64_dyn_b(v);
}

function unpack_u64_dyn_v2(buf, offset = 0) {
  return unpack_u64_dyn_b(buf, offset);
}

function pack_i64_dyn(v) {
  return pack_i64_dyn_a(v);
}

function unpack_i64_dyn(buf, offset = 0) {
  return unpack_i64_dyn_a(buf, offset);
}

function pack_i64_dyn_v2(v) {
  return pack_i64_dyn_b(v);
}

function unpack_i64_dyn_v2(buf, offset = 0) {
  return unpack_i64_dyn_b(buf, offset);
}

module.exports = {
  pack_u64_dyn,
  unpack_u64_dyn,
  pack_u64_dyn_b,
  unpack_u64_dyn_b,
  pack_u64_dyn_p,
  unpack_u64_dyn_p,
  pack_u64_dyn_bp,
  unpack_u64_dyn_bp,
  pack_i64_dyn_a,
  unpack_i64_dyn_a,
  pack_i64_dyn_b,
  unpack_i64_dyn_b,
  pack_i64_dyn_bp,
  unpack_i64_dyn_bp,
  pack_u64_dyn_v2,
  unpack_u64_dyn_v2,
  pack_i64_dyn,
  unpack_i64_dyn,
  pack_i64_dyn_v2,
  unpack_i64_dyn_v2,
};

File: c/u64_dyn.c
#include "u64_dyn.h"

#if defined(_MSC_VER)
#include <intrin.h>
#endif

size_t pack_u64_dyn(uint8_t out[9], uint64_t v) {
  size_t i = 0;
  while (i != 8) {
    const uint8_t cur = v & 0x7f;
    v >>= 7;
    if (v != 0) {
      out[i++] = cur | 0x80;
    } else {
      out[i++] = cur;
      return i;
    }
  }
  out[i++] = v;
  return i;
}

bool unpack_u64_dyn(const uint8_t *in_data, size_t in_size, uint64_t *out_v,
                    size_t *out_size) {
  uint64_t v = 0;
  int bits = 0;
  size_t used = 0;
  while (1) {
    if (used >= in_size) {
      return false;
    }
    uint8_t b = in_data[used++];
    if (used == 9) {
      v += (uint64_t)b << bits;
      break;
    }
    v += (uint64_t)(b & 0x7f) << bits;
    if ((b & 0x80) == 0) {
      break;
    }
    bits += 7;
  }
  if (out_size) {
    *out_size = used;
  }
  if (out_v) {
    *out_v = v;
  }
  return true;
}

size_t pack_u64_dyn_b(uint8_t out[9], uint64_t v) {
  size_t i = 0;
  while (i != 8) {
    const uint8_t cur = v & 0x7f;
    v >>= 7;
    if (v != 0) {
      out[i++] = cur | 0x80;
      v -= 1; // v2
    } else {
      out[i++] = cur;
      return i;
    }
  }
  out[i++] = v;
  return i;
}

bool unpack_u64_dyn_b(const uint8_t *in_data, size_t in_size, uint64_t *out_v,
                      size_t *out_size) {
  uint64_t v = 0;
  int bits = 0;
  size_t used = 0;
  while (1) {
    if (used >= in_size) {
      return false;
    }
    uint8_t b = in_data[used++];
    if (used == 9) {
      v += (uint64_t)b << bits;
      break;
    }
    v += (uint64_t)(b & 0x7f) << bits;
    if ((b & 0x80) == 0) {
      break;
    }
    bits += 7;
    v += (uint64_t)1 << bits; // v2
  }
  if (out_size) {
    *out_size = used;
  }
  if (out_v) {
    *out_v = v;
  }
  return true;
}

size_t pack_i64_dyn_a(uint8_t out[9], int64_t v) {
  uint64_t u = (uint64_t)v;
  uint64_t neg = (uint64_t)(v < 0);
  if (neg) {
    u = (~u + 1) & ~((uint64_t)1 << 63);
  }
  return pack_u64_dyn(out, (neg << 6) | ((u & ~0x3f) << 1) | (u & 0x3f));
}

bool unpack_i64_dyn_a(const uint8_t *in_data, size_t in_size, int64_t *out_v,
                      size_t *out_size) {
  uint64_t u;
  if (!unpack_u64_dyn(in_data, in_size, &u, out_size)) {
    return false;
  }
  const bool neg = (u >> 6) & 1;
  int64_t v = ((u >> 1) & ~((uint64_t)0x3f)) | (u & 0x3f);
  if (neg) {
    v = (int64_t)(~(v - 1) | ((uint64_t)1 << 63));
  }
  if (out_v) {
    *out_v = v;
  }
  return true;
}

size_t pack_i64_dyn_b(uint8_t out[9], int64_t v) {
  uint64_t neg = (uint64_t)(v < 0);
  uint64_t u = v ^ (0xffffffffffffffff * neg);
  return pack_u64_dyn_b(out,
                        (neg << 6) | ((u & ~0x3fULL) << 1) | (u & 0x3fULL));
}

bool unpack_i64_dyn_b(const uint8_t *in_data, size_t in_size, int64_t *out_v,
                      size_t *out_size) {
  uint64_t u;
  if (!unpack_u64_dyn_b(in_data, in_size, &u, out_size)) {
    return false;
  }
  const uint64_t neg = (u >> 6) & 1;         // check bit 6
  u = ((u >> 1) & ~0x3fULL) | (u & 0x3fULL); // remove bit 6
  int64_t v = u ^ (0xffffffffffffffff * neg);
  if (out_v) {
    *out_v = v;
  }
  return true;
}

size_t pack_u64_dyn_p(uint8_t out[9], uint64_t v) {
  const size_t byte_length = 1 + (v >= 1ull << 7) + (v >= 1ull << 14) +
                             (v >= 1ull << 21) + (v >= 1ull << 28) +
                             (v >= 1ull << 35) + (v >= 1ull << 42) +
                             (v >= 1ull << 49) + (v >= 1ull << 56);

  const size_t first_byte_value_bits = (byte_length < 8) * (8 - byte_length);
  const size_t first_byte_prefix_bits = byte_length - 1;

  out[0] = (0xff << (8 - first_byte_prefix_bits)) |
           (v & ((1 << first_byte_value_bits) - 1));
  v >>= first_byte_value_bits;
  out[1] = v & 0xff;
  out[2] = (v >> 8) & 0xff;
  out[3] = (v >> 16) & 0xff;
  out[4] = (v >> 24) & 0xff;
  out[5] = (v >> 32) & 0xff;
  out[6] = (v >> 40) & 0xff;
  out[7] = (v >> 48) & 0xff;
  out[8] = (v >> 56) & 0xff;
  return byte_length;
}

static unsigned int getNumLeadingZeros(uint32_t mask) {
#if defined(_MSC_VER)
  unsigned long ret;
  if (_BitScanReverse(&ret, mask)) {
    return 31 - ret;
  }
#else
  if (mask != 0) {
    return __builtin_clz(mask);
  }
#endif
  return 32;
}

bool unpack_u64_dyn_p(const uint8_t *in_data, size_t in_size, uint64_t *out_v,
                      size_t *out_size) {
  if (!in_size) {
    return false;
  }
  const size_t byte_length =
      1 + (getNumLeadingZeros((uint32_t)((uint8_t)~in_data[0])) - 24);
  const size_t first_byte_value_bits = (byte_length < 8) * (8 - byte_length);
  if (byte_length > in_size) {
    return false;
  }
  uint64_t v = 0;
  for (size_t i = 1; i != byte_length; ++i) {
    v |= (uint64_t)in_data[i] << ((i - 1) * 8);
  }
  v <<= first_byte_value_bits;
  v |= (in_data[0] & ((1 << first_byte_value_bits) - 1));
  if (out_v) {
    *out_v = v;
  }
  if (out_size) {
    *out_size = byte_length;
  }
  return true;
}

size_t pack_u64_dyn_v2(uint8_t out[9], uint64_t v) {
  return pack_u64_dyn_b(out, v);
}

bool unpack_u64_dyn_v2(const uint8_t *in_data, size_t in_size, uint64_t *out_v,
                       size_t *out_size) {
  return unpack_u64_dyn_b(in_data, in_size, out_v, out_size);
}

size_t pack_i64_dyn(uint8_t out[9], int64_t v) {
  return pack_i64_dyn_a(out, v);
}

bool unpack_i64_dyn(const uint8_t *in_data, size_t in_size, int64_t *out_v,
                    size_t *out_size) {
  return unpack_i64_dyn_a(in_data, in_size, out_v, out_size);
}

size_t pack_i64_dyn_v2(uint8_t out[9], int64_t v) {
  return pack_i64_dyn_b(out, v);
}

bool unpack_i64_dyn_v2(const uint8_t *in_data, size_t in_size, int64_t *out_v,
                       size_t *out_size) {
  return unpack_i64_dyn_b(in_data, in_size, out_v, out_size);
}
